File: MF4C.m
[filename, path] = uigetfile('*.txt', 'Select Full Pol config file');
FolderName = path;


config_ID = fopen(strcat(FolderName,'\','config.txt'),'rb');
tline = fgetl(config_ID);
tline = fgetl(config_ID);
b = str2num(tline); %row
tline = fgetl(config_ID);
tline = fgetl(config_ID);
tline = fgetl(config_ID);
a = str2num(tline); %column
nrow = b;
ncol = a;

fileList = dir('*.bin');

folderName = strcat(FolderName,'\','T11.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T11 = fread(fileID,[a b],'float32');
T11 = T11';

folderName = strcat(FolderName,'\','T12_imag.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T12_imag = fread(fileID,[a b],'float32');
T12_imag = T12_imag';

folderName = strcat(FolderName,'\','T12_real.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T12_real = fread(fileID,[a b],'float32');
T12_real = T12_real';

T12 = complex(T12_real,T12_imag);
T21 = conj(T12);

folderName = strcat(FolderName,'\','T13_imag.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T13_imag = fread(fileID,[a b],'float32');
T13_imag = T13_imag';

folderName = strcat(FolderName,'\','T13_real.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T13_real = fread(fileID,[a b],'float32');
T13_real = T13_real';

T13 = complex(T13_real,T13_imag);
T31 = conj(T13);

folderName = strcat(FolderName,'\','T22.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T22 = fread(fileID,[a b],'float32');
T22 = T22';

folderName = strcat(FolderName,'\','T23_imag.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T23_imag = fread(fileID,[a b],'float32');
T23_imag = T23_imag';

folderName = strcat(FolderName,'\','T23_real.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T23_real = fread(fileID,[a b],'float32');
T23_real = T23_real';

T23 = complex(T23_real,T23_imag);
T32 = conj(T23);

folderName = strcat(FolderName,'\','T33.bin');
disp(folderName);
fileID = fopen(folderName,'rb');
T33 = fread(fileID,[a b],'float32');
T33 = T33';

pd_f = zeros(nrow, ncol); % full
pv_f = zeros(nrow, ncol); % full
ps_f = zeros(nrow, ncol); % full
pc_f = zeros(nrow, ncol); % full
theta_val_f = zeros(nrow, ncol); % full
tau_val_f = zeros(nrow, ncol); % full
dop_val_f = zeros(nrow, ncol); % full
theta_dbl_f = zeros(nrow, ncol); % theta_dbl


disp('load complete');

%%
wsi = 7;
disp('Taking window size as 7')
wsj = wsi; % Number of columns in the window

inci=fix(wsi/2); % Up & down movement margin from the central row
incj=fix(wsj/2); % Left & right movement from the central column

starti=fix(wsi/2)+1; % Starting row for window processing
startj=fix(wsj/2)+1; % Starting column for window processing

stopj= nrow-inci; % Stop row for window processing
stopi= ncol-incj; % Stop column for window processing

%%
for ii = startj:stopj
  for jj = starti:stopi
    
    
        t11s = T11(ii-inci:ii+inci,jj-incj:jj+incj);
        t11s = mean(t11s(:));
        t12s = T12(ii-inci:ii+inci,jj-incj:jj+incj);
        t12s = mean(t12s(:));
        t21s = T21(ii-inci:ii+inci,jj-incj:jj+incj);
        t21s = mean(t21s(:));
        t13s = T13(ii-inci:ii+inci,jj-incj:jj+incj);
        t13s = mean(t13s(:));
        t31s = T31(ii-inci:ii+inci,jj-incj:jj+incj);
        t31s = mean(t31s(:));
        t22s = T22(ii-inci:ii+inci,jj-incj:jj+incj);
        t22s = mean(t22s(:));
        t23s = T23(ii-inci:ii+inci,jj-incj:jj+incj);
        t23s = mean(t23s(:));
        t32s = T32(ii-inci:ii+inci,jj-incj:jj+incj);
        t32s = mean(t32s(:));
        t33s = T33(ii-inci:ii+inci,jj-incj:jj+incj);
        t33s = mean(t33s(:));

        T = [t11s, t12s, t13s; t21s, t22s, t23s; t31s, t32s, t33s];

        dop_f = real(sqrt(1-(27*(det(T)./(trace(T).^3)))));
        dop_val_f(ii,jj) = dop_f;

        k11_f = (t11s + t22s + t33s)/2;
        k44_f = (-t11s + t22s + t33s)/2;
        k14_f = imag(t23s);
        
        

        s0_f = trace(T);

        
        val1 = (4*dop_f*k11_f*k44_f)./(k44_f^2 - (1 + 4*dop_f^2)*k11_f^2);
        val2 = abs(k14_f)./(k11_f);
       
        theta_f = atand(val1);
        tau_f = atand(val2);
        theta_val_f(ii,jj) = theta_f;
        tau_val_f(ii,jj) = tau_f;
         
        pc_f(ii,jj) = dop_f.*s0_f.*sind(2.*tau_f);
        pv_f(ii,jj) = (1-dop_f).*s0_f;
        res_pow = s0_f - (pc_f(ii,jj) + pv_f(ii,jj));
        ps_f(ii,jj) = (res_pow/2).*(1+sind(2*theta_f));
        pd_f(ii,jj) = (res_pow/2).*(1-sind(2*theta_f));
  end
  fprintf('row: %d\n',ii);
end
%%
path = FolderName;
if ~exist(strcat(path,'MF4C'), 'dir')
    mkdir(strcat(path,'MF4C'));
end
fclose('all');
FolderName = path;
Fold = strcat(FolderName,'MF4C\');

path = Fold;
cd(path);
%%

fName = 'DoP_FP';
f_name_dopFP = strcat('\', char(fName), '.bin');
fileandpath_dopFP=strcat([path f_name_dopFP]);
fid_011 = fopen(fileandpath_dopFP,'wb');
fwrite(fid_011,dop_val_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('DoP_FP bin and hdr files written')

fName = 'theta_FP';
f_name_thetaFP = strcat('\', char(fName), '.bin');
fileandpath_thetaFP=strcat([path f_name_thetaFP]);
fid_01 = fopen(fileandpath_thetaFP,'wb');
fwrite(fid_01,theta_val_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('theta_FP bin and hdr files written');

fName = 'theta_dbl';
f_name_thetaDBL = strcat('\', char(fName), '.bin');
fileandpath_thetaDBL=strcat([path f_name_thetaDBL]);
fid_01 = fopen(fileandpath_thetaDBL,'wb');
fwrite(fid_01,theta_dbl_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('theta_dbl bin and hdr files written');

fName = 'tau_FP';
f_name_tauFP = strcat('\', char(fName), '.bin');
fileandpath_tauFP=strcat([path f_name_tauFP]);
fid_01 = fopen(fileandpath_tauFP,'wb');
fwrite(fid_01,tau_val_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('tauFP bin and hdr files written');

fName = 'pd_FP';
f_name_pdFP = strcat('\', char(fName), '.bin');
fileandpath_pdFP=strcat([path f_name_pdFP]);
fid_04 = fopen(fileandpath_pdFP,'wb');
fwrite(fid_04,pd_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('pd_FP bin and hdr files written');

fName = 'pv_FP';
f_name_pvFP = strcat('\', char(fName), '.bin');
fileandpath_pvFP=strcat([path f_name_pvFP]);
fid_07 = fopen(fileandpath_pvFP,'wb');
fwrite(fid_07,pv_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('pv_FP bin and hdr files written');

fName = 'ps_FP';
f_name_psFP = strcat('\', char(fName), '.bin');
fileandpath_psFP=strcat([path f_name_psFP]);
fid_10 = fopen(fileandpath_psFP,'wb');
fwrite(fid_10,ps_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('ps_FP bin and hdr files written');

fName = 'pc_FP';
f_name_pcFP = strcat('\', char(fName), '.bin');
fileandpath_pcFP=strcat([path f_name_pcFP]);
fid_11 = fopen(fileandpath_pcFP,'wb');
fwrite(fid_11,pc_f', 'float32');
hdrwrite_envi(fName, path, nrow, ncol);
disp('pc_FP bin and hdr files written');

fclose('all');

File: MF4C.R
require(ggplot2)
require(ggthemes)
require(ggstance)
require(extrafont)
require(complexplus)
require(OpenImageR)
require(envi)
require(float)
library(caTools)
library(raster)
library(rgdal)
library(rasterVis)
library(RColorBrewer)

#font_import()
#loadfonts()
##---------------------------------------------------------------------------------------
readComplexFloat <- function(f, size, endian){
  X2 <- readBin(f, double(), n = 2*size, size = 4, endian = endian) # caso 4
  X = complex(length.out = size)
  
  for(i in 1:size) {
    twoi = 2*i
    X[i] = complex(real = X2[twoi-1], imaginary=X2[twoi])
  }
  
  return(X)
}

myread.ENVI <- function (filename, headerfile = paste(filename, ".hdr", sep = ""))
{
  nCol <- nRow <- nBand <- data.type <- header.offset <- byte.order <- (-1)
  interleave = "bsq"
  if (!file.exists(headerfile))
    stop("read.ENVI: Could not open input header file: ",
         headerfile)
  Lines = read.table(headerfile, sep = "=", strip.white = TRUE,
                     row.names = NULL, as.is = TRUE, fill = TRUE)
  Fields = c("samples", "lines", "bands", "data type", "header offset",
             "interleave", "byte order")
  for (i in 1:nrow(Lines)) {
    Lab = tolower(Lines[i, 1])
    Lab = gsub("[ ]+", " ", Lab)
    j = match(Lab, Fields)
    Val = Lines[i, 2]
    if (length(j) == 1)
      switch(j, nCol <- as.integer(Val), nRow <- as.integer(Val),
             nBand <- as.integer(Val), data.type <- as.integer(Val),
             header.offset <- as.integer(Val), interleave <- gsub(" ",
                                                                  "", Val), byte.order <- as.integer(Val))
  }
  if (nCol <= 0 | nRow <= 0 | nBand <= 0)
    stop("read.ENVI: Error in input header file ", headerfile,
         " data sizes missing or incorrect", nRow, nCol, nBand)
  if (!(data.type %in% c(1, 2, 3, 4, 5, 6, 9, 12)))
    stop("read.ENVI: Error in input header file ", headerfile,
         " data type is missing, incorrect or unsupported ")
  ieee = if (.Platform$endian == "big")
    1
  else 0
  endian = if (ieee == byte.order | byte.order < 0)
    .Platform$endian
  else "swap"
  size = nRow * nCol * nBand
  if (!file.exists(filename))
    stop("read.ENVI: Could not open input file: ", filename)
  f = file(filename, "rb")
  if (header.offset > 0)
    readBin(f, raw(), n = header.offset)
  switch(data.type,
         X <- readBin(f, integer(), n = size, size = 1, signed = FALSE), # caso 1
         X <- readBin(f, integer(), n = size, size = 2, endian = endian), # caso 2
         X <- readBin(f, integer(), n = size, endian = endian), # caso 3
         X <- readBin(f, double(), n = size, size = 4, endian = endian), # caso 4
         X <- readBin(f, double(), n = size, endian = endian), # caso 5
         X <- readComplexFloat(f, size = size, endian = endian), # caso 6
         ,  # caso 7
         ,  # caso 8
         X <- readBin(f, complex(), n = size, endian = endian), # caso 9
         , # caso 10
         , # caso 11
         X <- readBin(f, integer(), n = size, size = 2, endian = endian, signed = FALSE) # caso 12
  )
  close(f)
  Fields = c("bil", "bip", "bsq")
  j = match(interleave, Fields)
  if (length(j) == 0)
    stop("read.ENVI: Error in input header file ", headerfile,
         " incorrect interleave type")
  switch(j, {
    dim(X) <- c(nCol, nBand, nRow)
    X <- aperm(X, c(3, 1, 2))
  }, {
    dim(X) <- c(nBand, nCol, nRow)
    X <- aperm(X, c(3, 2, 1))
  }, {
    dim(X) <- c(nCol, nRow, nBand)
    X <- aperm(X, c(2, 1, 3))
  })
  if (nBand == 1)
    dim(X) = c(nRow, nCol)
  return(X)
}

para01 <- function(x) {
  valores <- range(x)
  y = (x - valores[1]) / (valores[2] - valores[1])
  y
}


##----------------------------------------------------------------------------
main_dir = 'D:\\MF4CF\\T3\\'

# Read coherence elements (T3)
T11 <- myread.ENVI(paste0(main_dir,"T11.bin"), 
                   paste0(main_dir,"T11.bin.hdr"))
T12_imag <- myread.ENVI(paste0(main_dir,"T12_imag.bin"), 
                        paste0(main_dir,"T12_imag.bin.hdr"))
T12_real <- myread.ENVI(paste0(main_dir,"T12_real.bin"), 
                        paste0(main_dir,"T12_real.bin.hdr"))

# T12 = T12_real + T12_imag*1i
# T21 = Conj(T12)

T13_imag <- myread.ENVI(paste0(main_dir,"T13_imag.bin"), 
                        paste0(main_dir,"T13_imag.bin.hdr"))
T13_real <- myread.ENVI(paste0(main_dir,"T13_real.bin"), 
                        paste0(main_dir,"T13_real.bin.hdr"))

# T13 = T13_real + T13_imag*1i
# T31 = Conj(T13)

T22 <- myread.ENVI(paste0(main_dir,"T22.bin"), 
                   paste0(main_dir,"T22.bin.hdr"))
T23_imag <- myread.ENVI(paste0(main_dir,"T23_imag.bin"), 
                        paste0(main_dir,"T23_imag.bin.hdr"))
T23_real <- myread.ENVI(paste0(main_dir,"T23_real.bin"), 
                        paste0(main_dir,"T23_real.bin.hdr"))

# T23 = T23_real + T23_imag*1i
# T32 = Conj(T23)

T33 <- myread.ENVI(paste0(main_dir,"T33.bin"), 
                   paste0(main_dir,"T33.bin.hdr"))

Nrw = nrow(T11)
Ncl = ncol(T11)

##----------------------------------------------------------------------------
# selection of window
wsi = 7; # window size

kernel = matrix(1, nrow = wsi, ncol = wsi) / (wsi*wsi)


t11s = convolution(T11, kernel, "same")
T12_imag_filt = convolution(T12_imag, kernel, "same")
T12_real_filt = convolution(T12_real, kernel, "same")

t12s = T12_real_filt + T12_imag_filt*1i
t21s = Conj(t12s)

T13_imag_filt = convolution(T13_imag, kernel, "same")
T13_real_filt = convolution(T13_real, kernel, "same")

t13s = T13_real_filt + T13_imag_filt*1i
t31s = Conj(t13s)

t22s = convolution(T22, kernel, "same")

T23_imag_filt = convolution(T23_imag, kernel, "same")
T23_real_filt = convolution(T23_real, kernel, "same")

t23s = T23_real_filt + T23_imag_filt*1i
t32s = Conj(t23s)

t33s = convolution(T33, kernel, "same")

det_T3 = t11s*(t22s*t33s-t23s*t32s)-t12s*(t21s*t33s-t23s*t31s)+t13s*(t21s*t32s-t22s*t31s)
trace_T3 = t11s + t22s + t33s
m1 = Re(sqrt(1-(27*(det_T3/(trace_T3^3)))))

k11_f = (t11s + t22s + t33s)/2
k44_f = (-t11s + t22s + t33s)/2
k14_f = Im(t23s)

s0_f = trace_T3
dop_f = m1

val1 = (4*dop_f*k11_f*k44_f)/(k44_f^2 - (1 + 4*dop_f^2)*k11_f^2)
val2 = abs(k14_f)/(k11_f)

theta_f = atan(val1)*(180/pi) # separation for surface and dbl
tau_f = atan(val2)*(180/pi) # separation for helix

pc_f = dop_f*s0_f*(sin(2*tau_f*(pi/180)))
pv_f = (1-dop_f)*s0_f
res_pow = s0_f - (pc_f + pv_f)
ps_f = (res_pow/2)*(1+sin((2*theta_f*(pi/180))))
pd_f = (res_pow/2)*(1-sin((2*theta_f*(pi/180))))

#-----------------------------------------------------------------------------
# Save in envi format
dir.create(file.path(main_dir, 'MF4CF_Rext'), showWarnings = FALSE)
save_dir = paste0(main_dir,'MF4CF_Rext\\')

r <- raster(theta_f)
r <- writeRaster(r, filename=paste0(save_dir,'theta_f.envi'), overwrite=TRUE)

r <- raster(tau_f)
r <- writeRaster(r, filename=paste0(save_dir,'tau_f.envi'), overwrite=TRUE)

r <- raster(pc_f)
r <- writeRaster(r, filename=paste0(save_dir,'pc_f.envi'), overwrite=TRUE)

r <- raster(pv_f)
r <- writeRaster(r, filename=paste0(save_dir,'pv_f.envi'), overwrite=TRUE)

r <- raster(ps_f)
r <- writeRaster(r, filename=paste0(save_dir,'ps_f.envi'), overwrite=TRUE)

r <- raster(pd_f)
r <- writeRaster(r, filename=paste0(save_dir,'pd_f.envi'), overwrite=TRUE)